fuel-vm/fuzz/src/lib.rs

```diff
@@ -0,0 +1,166 @@
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct FuzzData {
+    pub program: Vec<u8>,
+    pub sub_program: Vec<u8>,
+    pub script_data: Vec<u8>,
+}
+
+pub fn encode(data: &FuzzData) -> Vec<u8> {
+    let seperator: Vec<_> = SEP.into();
+    let parts: [Vec<u8>; 5] = [
+        data.program.iter().copied().collect(),
+        seperator.clone(),
+        data.script_data.clone(),
+        seperator.clone(),
+        data.sub_program.iter().copied().collect(),
+    ];
+
+    let data: Vec<u8> = parts.iter().flatten().copied().collect();
+    data
+}
+
+fn split_by_separator(data: &[u8], separator: &[u8]) -> Vec<Range<usize>> {
+    let separator_len = separator.len();
+
+    let mut last: usize = 0;
+
+    let mut result: Vec<_> = data
+        .windows(separator_len)
+        .enumerate()
+        .filter_map(|(i, window)| {
+            if window == separator {
+                let option = Some(last..i);
+                last = i + separator_len;
+                return option;
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    result.push(last..data.len());
+
+    result
+}
+
+pub fn decode(data: &[u8]) -> Option<FuzzData> {
+    let x = split_by_separator(data, &SEP);
+
+    if x.len() != 3 {
+        return None;
+    }
+
+    let program = data[x[0].clone()].to_vec();
+    let sub_program = data[x[2].clone()].to_vec();
+    Some(FuzzData {
+        program,
+        script_data: data[x[1].clone()].to_vec(),
+        sub_program,
+    })
+}
+
+pub fn decode_instructions(bytes: &[u8]) -> Option<Vec<Instruction>> {
+    let instructions: Vec<_> = fuel_vm::fuel_asm::from_bytes(bytes.iter().cloned())
+        .flat_map(|i: Result<Instruction, InvalidOpcode>| i.ok())
+        .collect();
+    return Some(instructions);
+}
+
+pub struct ExecuteResult {
+    pub success: bool,
+    pub gas_used: u64,
+}
+
+pub fn execute(data: FuzzData) -> ExecuteResult {
+    let gas_limit = 1_000_000;
+    let asset_id: AssetId = AssetId::new([
+        0xFA, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+        1, 1, 1, 1, 1, 1,
+    ]);
+
+    let mut test_context = TestBuilder::new(2322u64);
+    let subcontract: Vec<_> = [
+        // Pass some data to fuzzer
+        op::addi(
+            0x10,
+            fuel_asm::RegId::FP,
+            CallFrame::b_offset() as Immediate12,
+        ),
+        fuel_vm::fuel_asm::op::lw(0x10, 0x10, 0), // load address word
+    ]
+    .iter()
+    .copied()
+    .flat_map(Instruction::to_bytes)
+    .chain(data.sub_program.iter().copied())
+    .collect::<Vec<u8>>();
+
+    let contract_id = test_context
+        .setup_contract_bytes(subcontract.clone(), None, None)
+        .contract_id;
+
+    let max_program_length = 2usize.pow(18)
+        - test_context.get_tx_params().tx_offset()
+        - <fuel_vm::prelude::Script as Script>::script_offset_static()
+        - 256; // TODO reevalute this one
```

Comment:
I'm not sure where this magic value comes from, but I guess it is oke to have that.